### PR TITLE
maestral: enable darwin support

### PR DESCRIPTION
**Pull request by natsukium**
###### Description of changes

- add `python3Packages.rubicon-objc` as a dependency
- enable darwin support

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 235871` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>maestral (python310Packages.maestral)</li>
    <li>maestral.dist (python310Packages.maestral.dist)</li>
    <li>python310Packages.desktop-notifier</li>
    <li>python310Packages.desktop-notifier.dist</li>
    <li>python310Packages.rubicon-objc</li>
    <li>python310Packages.rubicon-objc.dist</li>
    <li>python311Packages.desktop-notifier</li>
    <li>python311Packages.desktop-notifier.dist</li>
    <li>python311Packages.maestral</li>
    <li>python311Packages.maestral.dist</li>
    <li>python311Packages.rubicon-objc</li>
    <li>python311Packages.rubicon-objc.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/rubicon-objc/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    hash = "sha256-+qbjc96rk7pwiBNUuIQTe4BAPOePyiOA8YnLM4s1CSY=";
+  };
+
+  nativeBuildInputs = [
+    setuptools-scm
+  ];
+
+  propagatedBuildInputs = lib.optionals (pythonOlder "3.8") [
+    importlib-metadata
+  ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/rubicon-objc/default.nix:30]: We don't have anything older than 3.8 anymore.

**Review by mweinelt — DISMISSED** [review]

File: pkgs/development/python-modules/rubicon-objc/default.nix (hunk 1 above)
Comment (natsukium) [reply pkgs/development/python-modules/rubicon-objc/default.nix]: Thanks for the suggestion.
I removed unnecessary `propagatedBuildInputs`.

**Comment by wegank** [comment]
@ofborg build maestral maestral.passthru.tests

**Comment by SFrijters** [comment]
Result of `nixpkgs-review pr 235871` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.rubicon-objc</li>
    <li>python310Packages.rubicon-objc.dist</li>
    <li>python311Packages.rubicon-objc</li>
    <li>python311Packages.rubicon-objc.dist</li>
  </ul>
</details>

**Review by SFrijters — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 235871` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.rubicon-objc</li>
    <li>python310Packages.rubicon-objc.dist</li>
    <li>python311Packages.rubicon-objc</li>
    <li>python311Packages.rubicon-objc.dist</li>
  </ul>
</details>

I'm not sure why it tries to build those at all for x86_64-linux, but maybe rubicon-objc needs to have meta.platform set?

```
error: builder for '/nix/store/v8xg3x3a6qq8vvv59nm4fga40q61vzwp-python3.10-rubicon-objc-0.4.6.drv' failed with exit code 1;
       last 10 log lines:
       >   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
       >   File "/nix/store/4wr8i3qspiyn11l04r0hfs2h0r7vzl9k-python3.10-rubicon-objc-0.4.6/lib/python3.10/site-packages/rubicon/objc/__init__.py", line 33, in <module>
       >     from . import api, collections, runtime, types
       >   File "/nix/store/4wr8i3qspiyn11l04r0hfs2h0r7vzl9k-python3.10-rubicon-objc-0.4.6/lib/python3.10/site-packages/rubicon/objc/api.py", line 28, in <module>
       >     from .runtime import (
       >   File "/nix/store/4wr8i3qspiyn11l04r0hfs2h0r7vzl9k-python3.10-rubicon-objc-0.4.6/lib/python3.10/site-packages/rubicon/objc/runtime.py", line 115, in <module>
       >     libobjc = load_library("objc")
       >   File "/nix/store/4wr8i3qspiyn11l04r0hfs2h0r7vzl9k-python3.10-rubicon-objc-0.4.6/lib/python3.10/site-packages/rubicon/objc/runtime.py", line 111, in load_library
       >     raise ValueError(f"Library {name!r} not found")
       > ValueError: Library 'objc' not found
       For full logs, run 'nix log /nix/store/v8xg3x3a6qq8vvv59nm4fga40q61vzwp-python3.10-rubicon-objc-0.4.6.drv'.
error: builder for '/nix/store/aw8ym9b19xv8j19rbq5x0144g7z7wn60-python3.11-rubicon-objc-0.4.6.drv' failed with exit code 1;
       last 10 log lines:
       >   File "/nix/store/8zw1limsxxm51z2spl7589zzz4dix5ik-python3.11-rubicon-objc-0.4.6/lib/python3.11/site-packages/rubicon/objc/__init__.py", line 33, in <module>
       >     from . import api, collections, runtime, types
       >   File "/nix/store/8zw1limsxxm51z2spl7589zzz4dix5ik-python3.11-rubicon-objc-0.4.6/lib/python3.11/site-packages/rubicon/objc/api.py", line 28, in <module>
       >     from .runtime import (
       >   File "/nix/store/8zw1limsxxm51z2spl7589zzz4dix5ik-python3.11-rubicon-objc-0.4.6/lib/python3.11/site-packages/rubicon/objc/runtime.py", line 115, in <module>
       >     libobjc = load_library("objc")
       >               ^^^^^^^^^^^^^^^^^^^^
       >   File "/nix/store/8zw1limsxxm51z2spl7589zzz4dix5ik-python3.11-rubicon-objc-0.4.6/lib/python3.11/site-packages/rubicon/objc/runtime.py", line 111, in load_library
       >     raise ValueError(f"Library {name!r} not found")
       > ValueError: Library 'objc' not found
       For full logs, run 'nix log /nix/store/aw8ym9b19xv8j19rbq5x0144g7z7wn60-python3.11-rubicon-objc-0.4.6.drv'.
error: 2 dependencies of derivation '/nix/store/cpw48k3g78pkyabvgc778xm1m877iphq-review-shell.drv' failed to build
```

**Comment by natsukium** [comment]
@SFrijters san, thanks for your checking.
I've restricted the platform to dawin only.

**Comment by natsukium** [comment]
`passthru.tests` seems to be broken in the master as well.

**Review by SFrijters — APPROVED** [review]

**Comment by SFrijters** [comment]
@natsukium The nixosTest seems to work for me locally (just getting a trace warning):
```
$ nix-build . -A nixosTests.maestral
trace: warning: Module argument `nodes.cli.config` is deprecated. Use `nodes.cli` instead.
/nix/store/lfbd24ydkjksjw6ma7ic83xmqchxyflz-vm-test-run-maestral
```

File: pkgs/development/python-modules/rubicon-objc/default.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+, fetchPypi
+, pythonOlder
+, setuptools-scm
+, unittestCheckHook
+}:
+
+buildPythonPackage rec {
+  pname = "rubicon-objc";
+  version = "0.4.6";
+  format = "pyproject";
```
Comment (sikmir) [inline pkgs/development/python-modules/rubicon-objc/default.nix:14]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/rubicon-objc/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  };
+
+  nativeBuildInputs = [
+    setuptools-scm
+  ];
+
+  preCheck = ''
+    pushd tests/objc
+    make
+    popd
```
Comment (sikmir) [inline pkgs/development/python-modules/rubicon-objc/default.nix:30]:
```suggestion
    make -C tests/objc
```

File: pkgs/development/python-modules/rubicon-objc/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+, darwin
+, fetchPypi
+, pythonOlder
+, setuptools-scm
+, unittestCheckHook
+}:
+
+buildPythonPackage rec {
+  pname = "rubicon-objc";
+  version = "0.4.6";
```
Comment (sikmir) [inline pkgs/development/python-modules/rubicon-objc/default.nix:13]: 0.4.7 is out

File: pkgs/development/python-modules/rubicon-objc/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  disabled = pythonOlder "3.8";
+
+  src = fetchFromGitHub {
+    owner = "beeware";
+    repo = "rubicon-objc";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-aFKzLeVYn5u8hTEgXCum3XpZxI7C/Wql41jkWkCF0HQ=";
+  };
+
+  postPatch = ''
```
Comment (SFrijters) [inline pkgs/development/python-modules/rubicon-objc/default.nix:25]: Can this be https://nixos.org/manual/nixpkgs/stable/#using-pythonrelaxdepshook instead?

File: pkgs/development/python-modules/rubicon-objc/default.nix [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+, fetchFromGitHub
+, pythonOlder
+, setuptools
+, setuptools-scm
+, unittestCheckHook
+}:
+
+buildPythonPackage rec {
+  pname = "rubicon-objc";
+  version = "0.4.8";
```
Comment (SFrijters) [inline pkgs/development/python-modules/rubicon-objc/default.nix:13]: Can you update the commit message to match the version number?

**Comment by SFrijters** [comment]
I don't have a Darwin system so I cannot really test this, but apart from my two comments it looks like it should be safe to merge?

File: pkgs/development/python-modules/rubicon-objc/default.nix (hunk 5 above)
Comment (natsukium) [reply pkgs/development/python-modules/rubicon-objc/default.nix:25]: Unfortunately, we can use pythonRelaxDepsHook for runtime dependencies, but not for build-time dependencies.

File: pkgs/development/python-modules/rubicon-objc/default.nix (hunk 6 above)
Comment (natsukium) [reply pkgs/development/python-modules/rubicon-objc/default.nix:13]: Good catch, thanks!

**Comment by natsukium** [comment]
Thanks for your review!
This patch in watchdog causes the maestral test to fail.  But I don't know what it does.
I'll try to see if it works on the real machine.
https://github.com/NixOS/nixpkgs/blob/4cba8b53da471aea2ab2b0c1f30a81e7c451f4b6/pkgs/development/python-modules/watchdog/force-kqueue.patch#L9-L13

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 235871` at f7ff5236 run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built successfully:</summary>
  <ul>
    <li>maestral (python311Packages.maestral)</li>
    <li>python311Packages.desktop-notifier</li>
    <li>python311Packages.rubicon-objc</li>
    <li>python312Packages.desktop-notifier</li>
    <li>python312Packages.rubicon-objc</li>
  </ul>
</details>

**Comment by natsukium** [comment]
The CI results are weird, but I think we can merge this PR.
The program works fine on my machine.

File: pkgs/development/python-modules/maestral/default.nix [hunk 7]
```diff
@@ -90,6 +97,19 @@ buildPythonPackage rec {
     "test_locking_multiprocess"
     # OSError: [Errno 95] Operation not supported
     "test_move_preserves_xattrs"
+  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    # maetral daemon does not start but worked in real environment
+    "test_catching_non_ignored_events"
```
Comment (SFrijters) [inline pkgs/development/python-modules/maestral/default.nix:102]: Very tiniest of nits: `maetral` -> `maestral`

**Review by SFrijters — APPROVED** [review]

File: pkgs/development/python-modules/maestral/default.nix (hunk 7 above)
Comment (natsukium) [reply pkgs/development/python-modules/maestral/default.nix:102]: Ahhh, I'll fix it next time. Thanks.

**Pull request merged**